src/main/java/com/iexec/core/task/update/TaskUpdate.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2021 IEXEC BLOCKCHAIN TECH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.iexec.core.task.update;
+
+import com.iexec.core.task.Task;
+import com.iexec.core.task.TaskStatus;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Comparator;
+import java.util.Date;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentMap;
+
+@Slf4j
+class TaskUpdate implements Runnable, Comparable<TaskUpdate> {
+    private final Task task;
+    private final ConcurrentMap<String, Object> locks;
+    private final TaskUpdateRequestConsumer consumer;
+
+    TaskUpdate(Task task,
+               ConcurrentMap<String, Object> locks,
+               TaskUpdateRequestConsumer consumer) {
+        this.task = task;
+        this.locks = locks;
+        this.consumer = consumer;
+    }
+
+    public Task getTask() {
+        return task;
+    }
+
+    public String getChainTaskId() {
+        return task.getChainTaskId();
+    }
+
+    private TaskStatus getCurrentStatus() {
+        return task.getCurrentStatus();
+    }
+
+    private Date getContributionDeadline() {
+        return task.getContributionDeadline();
+    }
+
+    /**
+     * Updates a task.
+     * <br>
+     * 2 updates can be run in parallel if they don't target the same task.
+     * Otherwise, the second update will wait until the first one is achieved.
+     */
+    @Override
+    public void run() {
+        if (task == null) {
+            return;
+        }
+
+        String chainTaskId = task.getChainTaskId();
+        log.debug("Selected task [chainTaskId: {}, status: {}]", chainTaskId, task.getCurrentStatus());
+        synchronized (locks.computeIfAbsent(chainTaskId, key -> new Object())) { // require one update on a same task at a time
+            consumer.onTaskUpdateRequest(chainTaskId); // synchronously update task
+        }
+    }
+
+    @Override
+    public int compareTo(TaskUpdate otherTaskUpdate) {
```

Comment:
This or `null` check
```suggestion
    public int compareTo(@NotNull TaskUpdate otherTaskUpdate) {
```

Reply:
I agree for `null` check. `@NotNull` is an annotation provided by JetBrains and I'd prefer not to add some IDE-dependent dependency.

Reply:
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Comparator.html#compare(T,T)
Can throw `NullPointerException`, we just need to handle it the day we use `compareTo` (does not seem to be the case right now).

Reply:
It's used inside `TaskUpdatePriorityBlockingQueue` as the natural-ordering comparator.
Having a look at your link, throwing an NPE seems to be the thing to do:
```
compare(x, y) must throw an exception if and only if compare(y, x) throws an exception.
```
Let me update this!